Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,935 @@
+  obtain ⟨n, y, h⟩ := h x
+  exact ⟨n, (algebraMap F E) y, h.symm ▸ (IsScalarTower.algebraMap_apply F E K y).symm⟩
+
+/-- If `E / F` and `K / E` are both purely inseparable extensions, then `K / F` is also
+purely inseparable. -/
+theorem IsPurelyInseparable.trans [Algebra E K] [IsScalarTower F E K]
+    [h1 : IsPurelyInseparable F E] [h2 : IsPurelyInseparable E K] : IsPurelyInseparable F K := by
+  obtain ⟨q, _⟩ := ExpChar.exists F
+  haveI := expChar_of_injective_algebraMap (algebraMap F E).injective q
+  rw [isPurelyInseparable_iff_pow_mem _ q] at h1 h2 ⊢
+  intro x
+  obtain ⟨n, y, h2⟩ := h2 x
+  obtain ⟨m, z, h1⟩ := h1 y
+  refine ⟨n + m, z, ?_⟩
+  rw [IsScalarTower.algebraMap_apply F E K, h1, map_pow, h2, ← pow_mul, ← pow_add]
+
+variable {E}
+
+/-- A field extension `E / F` is purely inseparable if and only if for every element `x` of `E`,
+its minimal polynomial has separable degree one. -/
+theorem isPurelyInseparable_iff_natSepDegree_eq_one :
+    IsPurelyInseparable F E ↔ ∀ x : E, (minpoly F x).natSepDegree = 1 := by
+  obtain ⟨q, _⟩ := ExpChar.exists F
+  simp_rw [isPurelyInseparable_iff_pow_mem F q, minpoly.natSepDegree_eq_one_iff_pow_mem q]
+
+theorem IsPurelyInseparable.natSepDegree_eq_one [IsPurelyInseparable F E] (x : E) :
+    (minpoly F x).natSepDegree = 1 :=
+  (isPurelyInseparable_iff_natSepDegree_eq_one F).1 ‹_› x
+
+/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
+if and only if for every element `x` of `E`, the minimal polynomial of `x` over `F` is of form
+`X ^ (q ^ n) - y` for some natural number `n` and some element `y` of `F`. -/
+theorem isPurelyInseparable_iff_minpoly_eq_X_pow_sub_C (q : ℕ) [hF : ExpChar F q] :
+    IsPurelyInseparable F E ↔ ∀ x : E, ∃ (n : ℕ) (y : F), minpoly F x = X ^ q ^ n - C y := by
+  simp_rw [isPurelyInseparable_iff_natSepDegree_eq_one,
+    minpoly.natSepDegree_eq_one_iff_eq_X_pow_sub_C q]
+
+theorem IsPurelyInseparable.minpoly_eq_X_pow_sub_C (q : ℕ) [ExpChar F q] [IsPurelyInseparable F E]
+    (x : E) : ∃ (n : ℕ) (y : F), minpoly F x = X ^ q ^ n - C y :=
+  (isPurelyInseparable_iff_minpoly_eq_X_pow_sub_C F q).1 ‹_› x
+
+/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
+if and only if for every element `x` of `E`, the minimal polynomial of `x` over `F` is of form
+`(X - x) ^ (q ^ n)` for some natural number `n`. -/
+theorem isPurelyInseparable_iff_minpoly_eq_X_sub_C_pow (q : ℕ) [hF : ExpChar F q] :
+    IsPurelyInseparable F E ↔
+    ∀ x : E, ∃ n : ℕ, (minpoly F x).map (algebraMap F E) = (X - C x) ^ q ^ n := by
+  simp_rw [isPurelyInseparable_iff_natSepDegree_eq_one,
+    minpoly.natSepDegree_eq_one_iff_eq_X_sub_C_pow q]
+
+theorem IsPurelyInseparable.minpoly_eq_X_sub_C_pow (q : ℕ) [ExpChar F q] [IsPurelyInseparable F E]
+    (x : E) : ∃ n : ℕ, (minpoly F x).map (algebraMap F E) = (X - C x) ^ q ^ n :=
+  (isPurelyInseparable_iff_minpoly_eq_X_sub_C_pow F q).1 ‹_› x
+
+variable (E)
+
+-- TODO: remove `halg` assumption
```

Comment:
What are we missing for this TODO?

Reply:
I think we need some result that for a transcendental extension `E / F`, the `finSepDegree` cannot be one, namely, there are at least two `F`-embeddings from `E` to the algebraic closure of `E`. Maybe @alreadydone has some preliminary works on it.

Reply:
Let's keep this for another PR, has this one is already pretty long.

Reply:
Mainly we need a version of [IsTranscendenceBasis.isAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/AlgebraicIndependent.html#IsTranscendenceBasis.isAlgebraic) with IntermediateField.adjoin rather than Algebra.adjoin, and we need to construct different embeddings of these IntermediateField.adjoin (which are purely transcendental extensions, and for this purpose IsTranscendenceBasis can be relaxed to AlgebraicIndependent.
If we want to count the number of embeddings, [even more work need to be done](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Separability.20degree.3F/near/411585334).